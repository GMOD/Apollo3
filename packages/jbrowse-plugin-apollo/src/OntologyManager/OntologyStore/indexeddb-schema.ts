/** schema types used to strongly-type using the `idb` type system */
import { DBSchema } from 'idb/with-async-ittr'

import {
  Edge as OboGraphEdge,
  Meta as OboGraphMeta,
  Node as OboGraphNode,
} from './obo-graph-json-schema'
import { SourceLocation } from '.'

/** metadata about this IndexedDB ontology database */
export interface Meta {
  /** original OntologyManager record this was loaded from */
  ontologyRecord: {
    name: string
    version: string
    sourceLocation: SourceLocation
  }
  /** graph metadata in OBO Graph metadata format */
  graphMeta?: OboGraphMeta
  timestamp: string
  /** IndexedDB schemaVersion for this ontology */
  schemaVersion: number
  /** time taken to load this data */
  timings: {
    /** milliseconds to fetch, parse, and load the ontology */
    overall: number
    /** optional milliseconds to fetch the ontology source file over the network */
    fetch?: number
    /** optional milliseconds to parse the ontology source file */
    parse?: number
    /** optional milliseconds to load the ontology into IndexedDB */
    load?: number
  }
}

// a Node that goes in the DB must have id and type, so make a new type that specifies this
export type OntologyDBNode = OboGraphNode & {
  id: string
  type: 'CLASS' | 'INDIVIDUAL' | 'PROPERTY'
}
export function isOntologyDBNode(node: OboGraphNode): node is OntologyDBNode {
  return typeof node.id === 'string'
}

// an Edge that goes in our DB must have sub, pred, and obj, so make a new type and guard that specifies this
export type OntologyDBEdge = OboGraphEdge & {
  sub: string
  pred: string
  obj: string
}
export function isOntologyDBEdge(edge: OboGraphEdge): edge is OntologyDBEdge {
  return (
    typeof edge.sub === 'string' &&
    typeof edge.pred === 'string' &&
    typeof edge.obj === 'string'
  )
}

export function isDeprecated(thing: OntologyDBNode | OntologyDBEdge) {
  return Boolean(thing.meta?.deprecated)
}

/** schema types used to strongly-type using the `idb` type system */
export interface OntologyDB extends DBSchema {
  meta: {
    key: string
    value: Meta
  }
  nodes: {
    key: string
    value: OntologyDBNode
    indexes: {
      'by-label': string
      'by-type': string
      'by-synonym': string
    }
  }
  edges: {
    key: number
    value: OntologyDBEdge
    indexes: {
      'by-subject': string
      'by-object': string
      'by-predicate': string
    }
  }
}
