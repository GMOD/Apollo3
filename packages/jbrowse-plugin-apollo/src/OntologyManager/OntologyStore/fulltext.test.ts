import { describe, expect, it } from '@jest/globals'

import {
  PREFIXED_ID_PATH,
  elaborateMatch,
  extractWords,
  getWords,
} from './fulltext'
import { OntologyDBNode } from './indexeddb-schema'

const testNode: OntologyDBNode = {
  id: 'http://purl.obolibrary.org/obo/SO_0000001',
  lbl: 'region',
  type: 'CLASS',
  meta: {
    definition: {
      val: 'A sequence_feature with an extent greater than zero. A nucleotide region is composed of bases and a polypeptide region is composed of amino acids. It may also be termed a region of some number of nucleotides.',
      xrefs: ['SO:ke'],
    },
    subsets: ['http://purl.obolibrary.org/obo/so#SOFA'],
    synonyms: [
      {
        pred: 'hasExactSynonym',
        val: 'sequence',
      },
    ],
    basicPropertyValues: [
      {
        pred: 'http://www.geneontology.org/formats/oboInOwl#hasOBONamespace',
        val: 'sequence',
      },
    ],
  },
}

const prefixes = new Map<string, string>([
  ['SO:', 'http://purl.obolibrary.org/obo/SO_'],
])

describe('extractWords', () => {
  it('can words from the members of objects', () => {
    const result = extractWords(['bar baz', 'noggin'])
    expect(Array.from(result)).toEqual(['bar', 'baz', 'noggin'])
  })
  it('can get the words from mix of stuff', () => {
    const set = extractWords(['zoz-zoo', 'bar baz', 'noggin', 'twenty'])
    expect(Array.from(set)).toEqual([
      'zoz',
      'zoo',
      'bar',
      'baz',
      'noggin',
      'twenty',
    ])
  })
})

describe('getWords', () => {
  it('can get the words from a test node', () => {
    const result = getWords(
      testNode,
      [
        PREFIXED_ID_PATH,
        '$.lbl',
        '$.meta.definition.val',
        '$.meta.synonyms[*].val',
      ],
      prefixes,
    )
    expect([...result]).toMatchSnapshot()
  })
})

describe('elaborateMatch', () => {
  it('can do one', () => {
    const result = elaborateMatch(
      [
        { displayName: 'label', jsonPath: '$.lbl' },
        { displayName: 'synonym', jsonPath: '$.meta.synonyms[*].val' },
        { displayName: 'definition', jsonPath: '$.meta.definition.val' },
      ],
      testNode,
      new Set([1, 2]),
      ['zonk', 'nucleotide', 'region'],
      prefixes,
    )
    expect(result.length).toBe(1)
    expect(result).toMatchSnapshot()
  })
  it('can do another', () => {
    const result = elaborateMatch(
      [
        { displayName: 'ID', jsonPath: PREFIXED_ID_PATH },
        { displayName: 'label', jsonPath: '$.lbl' },
        { displayName: 'synonym', jsonPath: '$.meta.synonyms[*].val' },
        { displayName: 'definition', jsonPath: '$.meta.definition.val' },
      ],
      testNode,
      new Set([0]),
      ['SO:0000001'],
      prefixes,
    )
    expect(result.length).toBe(1)
    expect(result).toMatchSnapshot()
  })
})
