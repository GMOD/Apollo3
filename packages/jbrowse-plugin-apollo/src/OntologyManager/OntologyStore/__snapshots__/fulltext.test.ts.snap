// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elaborateMatch can do another 1`] = `
[
  {
    "field": {
      "displayName": "ID",
      "jsonPath": "$PREFIXED_ID",
    },
    "score": 109.1,
    "str": "SO:0000001",
    "term": {
      "id": "http://purl.obolibrary.org/obo/SO_0000001",
      "lbl": "region",
      "meta": {
        "basicPropertyValues": [
          {
            "pred": "http://www.geneontology.org/formats/oboInOwl#hasOBONamespace",
            "val": "sequence",
          },
        ],
        "definition": {
          "val": "A sequence_feature with an extent greater than zero. A nucleotide region is composed of bases and a polypeptide region is composed of amino acids. It may also be termed a region of some number of nucleotides.",
          "xrefs": [
            "SO:ke",
          ],
        },
        "subsets": [
          "http://purl.obolibrary.org/obo/so#SOFA",
        ],
        "synonyms": [
          {
            "pred": "hasExactSynonym",
            "val": "sequence",
          },
        ],
      },
      "type": "CLASS",
    },
    "wordMatches": [
      {
        "position": 0,
        "wordIndex": 0,
      },
    ],
  },
]
`;

exports[`elaborateMatch can do one 1`] = `
[
  {
    "field": {
      "displayName": "definition",
      "jsonPath": "$.meta.definition.val",
    },
    "score": 208.39346153846154,
    "str": "A sequence_feature with an extent greater than zero. A nucleotide region is composed of bases and a polypeptide region is composed of amino acids. It may also be termed a region of some number of nucleotides.",
    "term": {
      "id": "http://purl.obolibrary.org/obo/SO_0000001",
      "lbl": "region",
      "meta": {
        "basicPropertyValues": [
          {
            "pred": "http://www.geneontology.org/formats/oboInOwl#hasOBONamespace",
            "val": "sequence",
          },
        ],
        "definition": {
          "val": "A sequence_feature with an extent greater than zero. A nucleotide region is composed of bases and a polypeptide region is composed of amino acids. It may also be termed a region of some number of nucleotides.",
          "xrefs": [
            "SO:ke",
          ],
        },
        "subsets": [
          "http://purl.obolibrary.org/obo/so#SOFA",
        ],
        "synonyms": [
          {
            "pred": "hasExactSynonym",
            "val": "sequence",
          },
        ],
      },
      "type": "CLASS",
    },
    "wordMatches": [
      {
        "position": 55,
        "wordIndex": 0,
      },
      {
        "position": 66,
        "wordIndex": 1,
      },
      {
        "position": 112,
        "wordIndex": 1,
      },
      {
        "position": 171,
        "wordIndex": 1,
      },
      {
        "position": 196,
        "wordIndex": 0,
      },
    ],
  },
]
`;

exports[`getWords can get the words from a test node 1`] = `
[
  [
    "$PREFIXED_ID",
    "so:0000001",
  ],
  [
    "$.lbl",
    "region",
  ],
  [
    "$.meta.definition.val",
    "sequence",
  ],
  [
    "$.meta.definition.val",
    "feature",
  ],
  [
    "$.meta.definition.val",
    "extent",
  ],
  [
    "$.meta.definition.val",
    "greater",
  ],
  [
    "$.meta.definition.val",
    "zero",
  ],
  [
    "$.meta.definition.val",
    "nucleotide",
  ],
  [
    "$.meta.definition.val",
    "region",
  ],
  [
    "$.meta.definition.val",
    "composed",
  ],
  [
    "$.meta.definition.val",
    "bases",
  ],
  [
    "$.meta.definition.val",
    "polypeptide",
  ],
  [
    "$.meta.definition.val",
    "region",
  ],
  [
    "$.meta.definition.val",
    "composed",
  ],
  [
    "$.meta.definition.val",
    "amino",
  ],
  [
    "$.meta.definition.val",
    "acids",
  ],
  [
    "$.meta.definition.val",
    "may",
  ],
  [
    "$.meta.definition.val",
    "also",
  ],
  [
    "$.meta.definition.val",
    "termed",
  ],
  [
    "$.meta.definition.val",
    "region",
  ],
  [
    "$.meta.definition.val",
    "number",
  ],
  [
    "$.meta.definition.val",
    "nucleotides",
  ],
  [
    "$.meta.synonyms[*].val",
    "sequence",
  ],
]
`;
